head.go

```diff
@@ -376,8 +378,9 @@ func (h *Head) loadWAL(r *wal.Reader) error {
 				firstInput <- append(buf[:0], samples[:n]...)
 				samples = samples[n:]
 			}
+			samples = s // Keep whole slice for reuse.
```

Comment:
Why did you need to keep the original slice here? After you'd fixed `samples` to be local to the whole function (the change on L359), didn't it be enough to fix the issue? I believe, `samples` and `s` are backed by the same array.

Reply:
They're backed by the same array, but samples would pointing at the end of it so it'd end up as a memory leak.